ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningCache.scala

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.interning
+
+import com.daml.dec.DirectExecutionContext
+import com.daml.lf.data.Ref
+import com.daml.logging.LoggingContext
+
+import scala.concurrent.Future
+
+class DomainStringIterators(
+    val parties: Iterator[String],
+    val templateIds: Iterator[String],
+)
+
+trait StringInterningCache extends StringInterning {
+
+  /** Fill StringInterning Cache with entries
+    *
+    * @param domainStringIterators iterators of the new entires
+    * @return If some of the entries were not part of the cache: they will be added, and these will be returned as a interned-id and raw, prefixed string pairs.
+    */
+  def add(domainStringIterators: DomainStringIterators): Iterable[(Int, String)]
+
+  /** Update the StringInterning Cache from persistence
+    *
+    * @param lastStringInterningId this is the "version" of the persistent cache, which from the StringInterningCache can see if it is behind
+    * @return a completion Future: if the cache is behind it will load the missing entries from persistence, and update the cache state
+    */
+  def updateStringInterningCache(lastStringInterningId: Int)(implicit
```

Comment:
This cache is very smart. For better separation of concerns, I'd like a dumber version of it: 
* Only has one method that updates the string interning data.
* An external loader takes care of updating from persistence.
* Have a `StringInterningLayer` that wires them up (for means of comprehensive testing - if we want to).
wdyt? In this manner, testing would also be greatly simplified

Reply:
I am not so sure about that: for example If you expose one method to update the string interning data, then also you have to expose the last id, and one of the maps in the Raw datastructure...and this is just the beginning ;)
I have the feeling that this trait represents perfectly the `StringInterningLayer` you suggest...
can we have a ppsession together where we explore your suggestion?

Reply:
As discussed offline we get back to this later